File: java/Order.java
import javax.crypto.Mac;
import java.net.URI;
import javax.crypto.spec.SecretKeySpec;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class Order {
    public static void main(String[] args) throws Exception {

        final String key = ""; // put your lalamove API key here
        final String secret = ""; // put your lalamove API secret here
        final String baseURL = "https://rest.sandbox.lalamove.com"; // URl to Lalamove Sandbox API
        long timestamp = System.currentTimeMillis();
        String method = "POST";
        String path = "/v2/orders";

        String body = "{" +
        "    \"serviceType\": \"MOTORCYCLE\"," +
        "    \"specialRequests\": []," +
        "    \"requesterContact\": {" +
        "        \"name\": \"test\"," +
        "        \"phone\": \"0899183138\"" +
        "    }," +
        "    \"stops\": [" +
        "        {" +
        "            \"location\": {" +
        "                \"lat\": \"1.3140113\"," +
        "                \"lng\": \"103.8807331\"" +
        "            }," +
        "            \"addresses\": {" +
        "                \"en_SG\": {" +
        "                    \"displayString\": \"Lorong 23 Geylang, Singapore Badminton Hall, Singapore\"," +
        "                    \"market\": \"SG_SIN\"" +
        "                }" +
        "            }" +
        "        }," +
        "        {" +
        "            \"location\": {" +
        "                \"lat\": \"1.2966147\"," +
        "                \"lng\": \"103.8485095\"" +
        "            }," +
        "            \"addresses\": {" +
        "                \"en_SG\": {" +
        "                    \"displayString\": \"Stamford Road, National Museum of Singapore, Singapore\"," +
        "                    \"market\": \"SG_SIN\"" +
        "                }" +
        "           }" +
        "        }" +
        "   ]," +
        "   \"deliveries\": [" +
        "        {" +
        "            \"toStop\": 1," +
        "            \"toContact\": {" +
        "                \"name\": \"dodo\"," +
        "                \"phone\": \"+660923447537\"" +
        "            }," +
        "           \"remarks\": \"Do not take this order - SANDBOX CLIENT TEST\"" +
        "        }" +
        "    ]," +
        "    \"quotedTotalFee\": {" +
        "        \"amount\": \"11.80\"," + // this is value from Quotation response, update it if needed
        "        \"currency\": \"SGD\"" +
        "    }" +
        "}";

        SecretKeySpec secret_key = new SecretKeySpec(secret.getBytes(), "HmacSHA256");
        Mac hmac256 = Mac.getInstance("HmacSHA256");
        hmac256.init(secret_key);

        String rawSignature = timestamp + "\r\n" + method + "\r\n" + path + "\r\n\r\n" + body;
        byte[] signatureRaw = hmac256.doFinal(rawSignature.getBytes("UTF-8"));
        StringBuilder buf = new StringBuilder();
        for (byte item : signatureRaw) {
            buf.append(Integer.toHexString((item & 0xFF) | 0x100).substring(1, 3));
        }

        long startTime = System.currentTimeMillis();
        String token = key + ':' + timestamp + ':' + buf.toString().toLowerCase();
        HttpClient client = HttpClient.newHttpClient();

        HttpRequest request = HttpRequest.newBuilder()
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .uri(URI.create(baseURL + path))
            .header("Content-Type", "application/json; charset=utf-8")
            .header("Accept", "application/json")
            .header("Authorization", "hmac " + token) // A unique Signature Hash has to be generated for EVERY API call at the time of making such call.
            .header("X-LLM-Market", "SG_SIN") // Please note to which city are you trying to make API call
            .build();

        HttpResponse<String> response = null;
        response = client.send(request, HttpResponse.BodyHandlers.ofString());

        long elapsedTime = System.currentTimeMillis() - startTime;
        System.out.println("Total elapsed http request/response time in milliseconds: " + elapsedTime);
        System.out.println("Authorization header: hmac " + token);
        System.out.println("Status code: " + response.statusCode());
        System.out.println("Response Body: " + response.body());
    }
}

File: c_sharp/order.cs
using System;
using System.Net;
using System.IO;
using System.Security.Cryptography;
using System.Text;

public class Order
{
    private static long ToUnixTimestamp(long time)
    {
        return (time - 621355968000000000) / TimeSpan.TicksPerMillisecond;
    }
    
    private static string HMACSHA256(string message, string key)
    {
        HMACSHA256 hmac = new HMACSHA256(Encoding.UTF8.GetBytes(key));
        hmac.Initialize();
        byte[] buffer = Encoding.UTF8.GetBytes(message);
        string signature = BitConverter.ToString(hmac.ComputeHash(buffer)).Replace("-", "").ToLower();
        return signature;
    }
    
    private static void PrintOutHttpResponse(WebResponse response, long startTime, string token)
    {
        HttpWebResponse httpResponse = (HttpWebResponse)response;
        Stream dataStream = response.GetResponseStream();
        StreamReader reader = new StreamReader(dataStream);
        string responseFromServer = reader.ReadToEnd();
        
        long endTime = ToUnixTimestamp(DateTime.Now.Ticks);
        long requestTime = endTime - startTime;
        
        // Get HTTP response
        Console.WriteLine("Total elapsed http request/response time in milliseconds: {0}", requestTime);
        Console.WriteLine("Authorization header: {0}", token);
        Console.WriteLine("Status Code: {0}", (int)httpResponse.StatusCode);
        Console.WriteLine("Returned data: {0}", responseFromServer);
        
        reader.Close();
        dataStream.Close();
        response.Close();
    }
    
    public static void Main(string[] args)
    {
        // Get information
        string key = "";    // put your lalamove API key here
        string secret = ""; // put your lalamove API secret here
        string baseUrl = "https://rest.sandbox.lalamove.com";
        string path = "/v2/orders";
        string region = "SG_SIN";
        string method = "POST";
        string time = ToUnixTimestamp(DateTime.Now.Ticks).ToString();
        string body = "{\"serviceType\": \"MOTORCYCLE\", \"specialRequests\": [], \"requesterContact\": {\"name\": \"test\", \"phone\": \"0899183138\"}, \"stops\": [{\"location\": {\"lat\": \"1.3140113\", \"lng\": \"103.8807331\"}, \"addresses\": {\"en_SG\": {\"displayString\": \"Lorong 23 Geylang, Singapore Badminton Hall, Singapore\", \"market\": \"SG_SIN\"}}}, {\"location\": {\"lat\": \"1.2966147\", \"lng\": \"103.8485095\"}, \"addresses\": {\"en_SG\": {\"displayString\": \"Stamford Road, National Museum of Singapore, Singapore\", \"market\": \"SG_SIN\"}}}], \"deliveries\": [{\"toStop\": 1, \"toContact\": {\"name\": \"dodo\", \"phone\": \"+660923447537\"}, \"remarks\": \"Do not take this order - SANDBOX CLIENT TEST\"}], \"quotedTotalFee\": {\"amount\": \"11.80\", \"currency\": \"SGD\"}}";
        
        // Create signature and authorization
        string rawSignature = String.Format("{0}\r\n{1}\r\n{2}\r\n\r\n{3}", time, method, path, body);
        string signature = HMACSHA256(rawSignature, secret);
        string token = string.Format("hmac {0}:{1}:{2}", key, time, signature);
        
        // Create HTTP request
        string url = baseUrl+path;
        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.Method = method;
        string postData = body;
        byte[] byteArray = Encoding.UTF8.GetBytes(postData);
        request.Accept = "application/json";
        request.ContentType = "application/json; charset=utf-8";
        request.Headers.Add("Authorization", token);
        request.Headers.Add("X-LLM-Market", region);
        
        long startTime = ToUnixTimestamp(DateTime.Now.Ticks);
        
        // Send HTTP request
        Stream dataStream = request.GetRequestStream();
        dataStream.Write(byteArray, 0, byteArray.Length);
        dataStream.Close();
        
        try 
        {
            WebResponse response = request.GetResponse();
            PrintOutHttpResponse(response, startTime, token);
        } 
        catch (WebException ex)
        {
            WebResponse response = ex.Response as WebResponse;
            PrintOutHttpResponse(response, startTime, token);
        }
    }
}